supervisor/exceptions.py

```diff
@@ -659,6 +659,10 @@ class BackupJobError(BackupError, JobException):
     """Raise on Backup job error."""
 
 
+class BackupFileNotFoundError(BackupError):
```

Comment:
```suggestion
class BackupFileNotFoundError(BackupError, APINotFound):
```
Then you can remove all the `try...except` logic added in `api/backups.py` converting `BackupFileNotFoundError` into `APINotFound`, that exception will just result in a 404 as is.

Reply:
Hm, that is a pattern we haven't used so far. Isn't that violating a bit separation of concerns here, as in the backup errors should not be dealing with API things? :thinking: Sure, this explicit catches add code, but it does seem cleaner to me.

Reply:
I don't really think it violates that personally. Things can have multiple inheritance in python and these are all basically just mixins. If an exception is expected to bubble up to the API and should always result in a particular status code it seems better to define it this way.

If anything we might want to consider auditing our other exception in a similar fashion. The only reason there isn't more examples of this right now is because Supervisor returned only 3 status codes:
- `200` for success
- `400` for handled error with a likely descriptive message (which meant the error extended from `HassioError` internally)
- `500` for unhandled error with an "unknown error, check logs" response (whcih meant the error did not extend from `HassioError` internally)

This isn't the situation anymore, Supervisor tries to use the status code to convey info about what went wrong. So more exceptions should be changed to translate to the appropriate code.

Reply:
Ok, but if we go down that road, then we should be consequent and mark the Supervisor Exception with the appropriate API error everywhere. E.g.

> * `400` for handled error with a likely descriptive message (which meant the error extended from `HassioError` internally)

Instead of special casing `HassioError` in API, we should make `HassioError` extend from `APIError`.

Or we go the other way around: Add the special exception in `api_process` (e.g. make `BackupFileNotFoundError` an `APINotFound` there).

I think I would prefer the latter.

Reply:
Well, it wouldn't really hurt if `HassioError` extends from `APIError` for these general cases.

But yeah, sort of that is my point: Any exception we define is potentially only used within Supervisor. Starting to mark some "random" exception how they should translate into API errors seems weird to me. What if we use and handle `BackupFileNotFoundError` somewhere not API related?

What do you think about handling `BackupFileNotFoundError` in `api_process` instead? :thinking:

Reply:
Sure that's fine. it just might turn into a long list. But that's fine, future problems, we can always refactor if we don't like it

Reply:
i realized this proposal was setting us up for a lot of `isinstance` checks to see what the type of error is. Those have notoriously poor performance so I refactored it to remove the need for them entirely from `api_return_error` instead of adding more. Let me know what you think.

Reply:
Uh nice, yeah I like that! :+1: